r/R/dataset.R

```diff
@@ -93,8 +93,11 @@ open_dataset <- function(sources,
     return(dataset___UnionDataset__create(sources, schema))
   }
   factory <- DatasetFactory$create(sources, partitioning = partitioning, ...)
-  # Default is _not_ to inspect/unify schemas
-  factory$Finish(schema, isTRUE(unify_schemas))
+  tryCatch(
+    # Default is _not_ to inspect/unify schemas
+    factory$Finish(schema, isTRUE(unify_schemas)),
+    error = handle_parquet_io_error
```

Comment:
One last thought: since we're in `open_dataset()` here, we know whether the user has specified a format or not. "Did you mean to specify a 'format'?" may be a good message if `format` is not specified, but maybe we should give a different message if you specified `format = "parquet"` (i.e., there is a non-parquet file in what you think is all parquet files).

Also maybe the default message should indicate that parquet is the default since that may not be obvious, something like "Did you mean to specify a 'format' other than the default (parquet)?" perhaps?

Reply:
I like this suggestion as it helps make the error message a lot more specific.  I've now updated the function so that it gives your updated error message (i.e. mentioning that parquet is the default) if the format is NULL (i.e. hasn't been specified by the user).  

I've tested the code with a directory containing a mix of file formats and the C++ error looks like this: 
`Error: IOError: Could not open parquet input source '/tmp/RtmpLi0d7E/filefa0e499bb459/file1.txt': Invalid: Parquet magic bytes not found in footer. Either the file is corrupted or this is not a parquet file.`

Given that the C++ error message above is pretty informative - it mentions both the name of the file which has caused the error and the likely source of the error - I haven't implemented anything that changes the error output in this case, as I think this is sufficient.